lib/Optimizer/Transforms/MemToReg.cpp

```diff
@@ -0,0 +1,689 @@
+    auto loc = op.getLoc();
+    SmallVector<Value> unwrapCtrls;
+    auto wireTy = quake::WireType::get(rewriter.getContext());
+    auto qrefTy = quake::RefType::get(rewriter.getContext());
+    // Scan the control and target positions. Any that were not Wires will be
+    // promoted to Wires via an unwrap operation. These unwrap ops become the
+    // arguments to the quantum value form of the new quantum operation.
+    if constexpr (!quake::isMeasure<OP>) {
+      for (auto opnd : op.getControls()) {
+        auto opndTy = opnd.getType();
+        if (opndTy == qrefTy) {
+          auto unwrap = rewriter.create<quake::UnwrapOp>(loc, wireTy, opnd);
+          unwrapCtrls.push_back(unwrap);
+        } else {
+          unwrapCtrls.push_back(opnd);
+        }
+      }
+    }
+    SmallVector<Value> unwrapTargs;
+    for (auto opnd : op.getTargets()) {
+      auto opndTy = opnd.getType();
+      if (opndTy == qrefTy) {
+        auto unwrap = rewriter.create<quake::UnwrapOp>(loc, wireTy, opnd);
+        unwrapTargs.push_back(unwrap);
+      } else {
+        unwrapTargs.push_back(opnd);
+      }
+    }
+    if constexpr (quake::isMeasure<OP>) {
+      // The result type is the same here.
+      rewriter.replaceOpWithNewOp<OP>(op, op.getType(), unwrapTargs,
+                                      op.getRegisterNameAttr());
+    } else {
+      // Scan the target positions. Any that were not Wires originally are now
+      // placed in the result vector. We propagate those new results to wrap
+      // operations.
+      auto newOp = rewriter.create<OP>(
+          loc, SmallVector<Type>{unwrapTargs.size(), wireTy}, op.getIsAdjAttr(),
+          op.getParameters(), unwrapCtrls, unwrapTargs,
+          op.getNegatedQubitControlsAttr());
+      for (auto i : llvm::enumerate(op.getTargets())) {
+        auto opndTy = i.value().getType();
+        unsigned count = 0;
+        if (opndTy == qrefTy) {
+          rewriter.create<quake::WrapOp>(loc, newOp.getResult(i.index()),
+                                         i.value());
+        } else if (opndTy == wireTy) {
+          op.getResult(count++).replaceAllUsesWith(newOp.getResult(i.index()));
+        }
+      }
+      rewriter.eraseOp(op);
+    }
+    return success();
+  }
+};
+
+static Type unwrapType(Type ty) {
+  if (isa<quake::RefType>(ty))
+    return quake::WireType::get(ty.getContext());
+  return cast<cudaq::cc::PointerType>(ty).getElementType();
+}
+
+#define WRAPPER(X) Wrapper<quake::X>
+#define WRAPPER_QUANTUM_OPS QUANTUM_OPS(WRAPPER)
+#define RAW(X) quake::X
+#define RAW_QUANTUM_OPS QUANTUM_OPS(RAW)
+
+namespace {
+class MemToRegPass : public cudaq::opt::impl::MemToRegBase<MemToRegPass> {
+public:
+  using MemToRegBase::MemToRegBase;
+  using DefnMap = DenseMap<Value, Value>;
+
+  void runOnOperation() override {
+    func::FuncOp func = getOperation();
+    LLVM_DEBUG(llvm::dbgs() << "Function before memtoreg:\n" << func << "\n\n");
```

Comment:
Can you remind me the command-line magic to turn these dbg statements on?

Reply:
--debug-only=<pass-debug-name>